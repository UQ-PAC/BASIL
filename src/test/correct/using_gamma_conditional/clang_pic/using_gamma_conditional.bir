000005de: program
000005aa: sub __cxa_finalize(__cxa_finalize_result)
000005df: __cxa_finalize_result :: out u32 = low:32[R0]

000003a9:
000004a1: R16 := 0x11000
000004a8: R17 := mem[R16 + 8, el]:u64
000004ae: R16 := R16 + 8
000004b3: call R17 with noreturn

000005ab: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005e0: __do_global_dtors_aux_result :: out u32 = low:32[R0]

00000292:
00000296: #3 := R31 - 0x20
0000029c: mem := mem with [#3, el]:u64 <- R29
000002a2: mem := mem with [#3 + 8, el]:u64 <- R30
000002a6: R31 := #3
000002ac: R29 := R31
000002b4: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002b9: R19 := 0x11000
000002c0: R0 := pad:64[mem[R19 + 0x30]]
000002c7: when 31:0[R0] <> 0 goto %000002c5
000005d3: goto %00000372

00000372:
00000375: R0 := 0x10000
0000037c: R0 := mem[R0 + 0xFC0, el]:u64
00000382: when R0 = 0 goto %00000380
000005d4: goto %00000399

00000399:
0000039c: R0 := 0x11000
000003a3: R0 := mem[R0 + 0x28, el]:u64
000003a8: R30 := 0x730
000003ab: call @__cxa_finalize with return %00000380

00000380:
00000388: R30 := 0x734
0000038a: call @deregister_tm_clones with return %0000038c

0000038c:
0000038f: R0 := 1
00000397: mem := mem with [R19 + 0x30] <- 7:0[R0]
000005d5: goto %000002c5

000002c5:
000002cf: R19 := mem[R31 + 0x10, el]:u64
000002d6: R29 := mem[R31, el]:u64
000002db: R30 := mem[R31 + 8, el]:u64
000002df: R31 := R31 + 0x20
000002e4: call R30 with noreturn

000005af: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005e1: __libc_start_main_main :: in u64 = R0
000005e2: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005e3: __libc_start_main_arg3 :: in out u64 = R2
000005e4: __libc_start_main_auxv :: in out u64 = R3
000005e5: __libc_start_main_result :: out u32 = low:32[R0]

000001eb:
0000048b: R16 := 0x11000
00000492: R17 := mem[R16, el]:u64
00000498: R16 := R16
0000049d: call R17 with noreturn

000005b0: sub _fini(_fini_result)
000005e6: _fini_result :: out u32 = low:32[R0]

0000001f:
00000025: #0 := R31 - 0x10
0000002b: mem := mem with [#0, el]:u64 <- R29
00000031: mem := mem with [#0 + 8, el]:u64 <- R30
00000035: R31 := #0
0000003b: R29 := R31
00000042: R29 := mem[R31, el]:u64
00000047: R30 := mem[R31 + 8, el]:u64
0000004b: R31 := R31 + 0x10
00000050: call R30 with noreturn

000005b1: sub _init(_init_result)
000005e7: _init_result :: out u32 = low:32[R0]

000004fb:
00000501: #6 := R31 - 0x10
00000507: mem := mem with [#6, el]:u64 <- R29
0000050d: mem := mem with [#6 + 8, el]:u64 <- R30
00000511: R31 := #6
00000517: R29 := R31
0000051c: R30 := 0x5A8
0000051e: call @call_weak_fn with return %00000520

00000520:
00000525: R29 := mem[R31, el]:u64
0000052a: R30 := mem[R31 + 8, el]:u64
0000052e: R31 := R31 + 0x10
00000533: call R30 with noreturn

000005b2: sub _start(_start_result)
000005e8: _start_result :: out u32 = low:32[R0]

000001ac:
000001b1: R29 := 0
000001b6: R30 := 0
000001bc: R5 := R0
000001c3: R1 := mem[R31, el]:u64
000001c9: R2 := R31 + 8
000001cf: R6 := R31
000001d4: R0 := 0x10000
000001db: R0 := mem[R0 + 0xFD8, el]:u64
000001e0: R3 := 0
000001e5: R4 := 0
000001ea: R30 := 0x670
000001ed: call @__libc_start_main with return %000001ef

000001ef:
000001f2: R30 := 0x674
000001f5: call @abort with return %000005d6

000005d6:
000005d7: call @call_weak_fn with noreturn

000005b5: sub abort()


000001f3:
000004cd: R16 := 0x11000
000004d4: R17 := mem[R16 + 0x18, el]:u64
000004da: R16 := R16 + 0x18
000004df: call R17 with noreturn

000005b6: sub call_weak_fn(call_weak_fn_result)
000005e9: call_weak_fn_result :: out u32 = low:32[R0]

000001f7:
000001fa: R0 := 0x10000
00000201: R0 := mem[R0 + 0xFD0, el]:u64
00000207: when R0 = 0 goto %00000205
000005d8: goto %000003e9

00000205:
0000020d: call R30 with noreturn

000003e9:
000003ec: goto @__gmon_start__

000003ea:
000004b7: R16 := 0x11000
000004be: R17 := mem[R16 + 0x10, el]:u64
000004c4: R16 := R16 + 0x10
000004c9: call R17 with noreturn

000005b8: sub deregister_tm_clones(deregister_tm_clones_result)
000005ea: deregister_tm_clones_result :: out u32 = low:32[R0]

00000213:
00000216: R0 := 0x11000
0000021c: R0 := R0 + 0x30
00000221: R1 := 0x11000
00000227: R1 := R1 + 0x30
0000022d: #1 := ~R0
00000232: #2 := R1 + ~R0
00000238: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000023e: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
00000242: ZF := #2 + 1 = 0
00000246: NF := 63:63[#2 + 1]
0000024c: when ZF goto %0000024a
000005d9: goto %000003cb

000003cb:
000003ce: R1 := 0x10000
000003d5: R1 := mem[R1 + 0xFB8, el]:u64
000003da: when R1 = 0 goto %0000024a
000005da: goto %000003de

0000024a:
00000252: call R30 with noreturn

000003de:
000003e2: R16 := R1
000003e7: call R16 with noreturn

000005bb: sub frame_dummy(frame_dummy_result)
000005eb: frame_dummy_result :: out u32 = low:32[R0]

000002ea:
000002ec: call @register_tm_clones with noreturn

000005bc: sub main(main_argc, main_argv, main_result)
000005ec: main_argc :: in u32 = low:32[R0]
000005ed: main_argv :: in out u64 = R1
000005ee: main_result :: out u32 = low:32[R0]

000002ee:
000002f2: R31 := R31 - 0x10
000002f9: mem := mem with [R31 + 0xC, el]:u32 <- 0
000002fe: R8 := 0x10000
00000305: R8 := mem[R8 + 0xFC8, el]:u64
0000030c: R8 := pad:64[mem[R8, el]:u32]
00000312: #4 := 31:0[R8] - 1
00000317: VF := extend:33[#4 + 1] <> extend:33[31:0[R8]] + 0
0000031c: CF := pad:33[#4 + 1] <> pad:33[31:0[R8]] - 0x100000000
00000320: ZF := #4 + 1 = 0
00000324: NF := 31:31[#4 + 1]
00000328: R8 := pad:64[#4 + 1]
00000334: when ZF <> 1 goto %0000032d
00000335: goto %00000330

00000330:
00000331: R8 := 1
00000337: goto %00000333

0000032d:
0000032e: R8 := 0
00000336: goto %00000333

00000333:
0000033d: when 0:0[R8] goto %0000033b
000005db: goto %00000363

0000033b:
00000343: R8 := 1
0000034b: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R8]
00000350: goto %0000034e

00000363:
00000366: goto %00000364

00000364:
0000036c: mem := mem with [R31 + 0xC, el]:u32 <- 0
00000370: goto %0000034e

0000034e:
00000356: R0 := pad:64[mem[R31 + 0xC, el]:u32]
0000035c: R31 := R31 + 0x10
00000361: call R30 with noreturn

000005be: sub register_tm_clones(register_tm_clones_result)
000005ef: register_tm_clones_result :: out u32 = low:32[R0]

00000254:
00000257: R0 := 0x11000
0000025d: R0 := R0 + 0x30
00000262: R1 := 0x11000
00000268: R1 := R1 + 0x30
0000026f: R1 := R1 + ~R0 + 1
00000275: R2 := 0.63:63[R1]
0000027c: R1 := R2 + (R1 ~>> 3)
00000282: R1 := extend:64[63:1[R1]]
00000288: when R1 = 0 goto %00000286
000005dc: goto %000003ad

000003ad:
000003b0: R2 := 0x10000
000003b7: R2 := mem[R2 + 0xFE0, el]:u64
000003bc: when R2 = 0 goto %00000286
000005dd: goto %000003c0

00000286:
0000028e: call R30 with noreturn

000003c0:
000003c4: R16 := R2
000003c9: call R16 with noreturn
