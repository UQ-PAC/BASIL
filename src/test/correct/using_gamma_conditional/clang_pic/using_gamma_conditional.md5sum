15b5d0f891a53c017b9a746c33686003  correct/using_gamma_conditional/clang_pic/a.out
0bc115baf544f4232d2c6ac5f14cf6ab  correct/using_gamma_conditional/clang_pic/using_gamma_conditional.adt
b885135350d6b921155b539b5fffef41  correct/using_gamma_conditional/clang_pic/using_gamma_conditional.bir
dbba58a029aa7b9d8e5665d68036a4b6  correct/using_gamma_conditional/clang_pic/using_gamma_conditional.relf
b58f9dd0e2f485fd6c126b3392a1760a  correct/using_gamma_conditional/clang_pic/using_gamma_conditional.gts
