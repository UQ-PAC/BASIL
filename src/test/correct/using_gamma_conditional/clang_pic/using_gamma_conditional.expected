var {:extern} CF: bv1;
var {:extern} Gamma_CF: bool;
var {:extern} Gamma_NF: bool;
var {:extern} Gamma_R0: bool;
var {:extern} Gamma_R31: bool;
var {:extern} Gamma_R8: bool;
var {:extern} Gamma_VF: bool;
var {:extern} Gamma_ZF: bool;
var {:extern} Gamma_mem: [bv64]bool;
var {:extern} Gamma_stack: [bv64]bool;
var {:extern} NF: bv1;
var {:extern} R0: bv64;
var {:extern} R31: bv64;
var {:extern} R8: bv64;
var {:extern} VF: bv1;
var {:extern} ZF: bv1;
var {:extern} mem: [bv64]bv8;
var {:extern} stack: [bv64]bv8;
const {:extern} $x_addr: bv64;
axiom ($x_addr == 69684bv64);
const {:extern} $z_addr: bv64;
axiom ($z_addr == 69688bv64);
function {:extern} L(memory: [bv64]bv8, index: bv64) returns (bool) {
  (if (index == $z_addr) then true else (if (index == $x_addr) then (memory_load32_le(memory, $z_addr) == 0bv32) else false))
}

function {:extern} {:bvbuiltin "bvadd"} bvadd32(bv32, bv32) returns (bv32);
function {:extern} {:bvbuiltin "bvadd"} bvadd33(bv33, bv33) returns (bv33);
function {:extern} {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp1(bv1, bv1) returns (bv1);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp32(bv32, bv32) returns (bv1);
function {:extern} {:bvbuiltin "bvcomp"} bvcomp33(bv33, bv33) returns (bv1);
function {:extern} {:bvbuiltin "bvnot"} bvnot1(bv1) returns (bv1);
function {:extern} gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function {:extern} gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function {:extern} gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function {:extern} memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function {:extern} memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function {:extern} memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function {:extern} memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:extern} {:bvbuiltin "sign_extend 1"} sign_extend1_32(bv32) returns (bv33);
function {:extern} {:bvbuiltin "zero_extend 1"} zero_extend1_32(bv32) returns (bv33);
function {:extern} {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
procedure {:extern} rely();
  modifies Gamma_mem, mem;
  ensures (forall i: bv64 :: (((mem[i] == old(mem[i])) ==> (Gamma_mem[i] == old(Gamma_mem[i])))));
  ensures (old(gamma_load32(Gamma_mem, $x_addr)) ==> gamma_load32(Gamma_mem, $x_addr));
  free ensures (memory_load8_le(mem, 1964bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1965bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1966bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1967bv64) == 0bv8);
  free ensures (memory_load64_le(mem, 69576bv64) == 69684bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free ensures (memory_load64_le(mem, 69056bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 69064bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1876bv64);

procedure {:extern} rely_transitive();
  modifies Gamma_mem, mem;
  ensures (old(gamma_load32(Gamma_mem, $x_addr)) ==> gamma_load32(Gamma_mem, $x_addr));

implementation {:extern} rely_transitive()
{
  call rely();
  call rely();
}

procedure {:extern} rely_reflexive();

implementation {:extern} rely_reflexive()
{
  assert (gamma_load32(Gamma_mem, $x_addr) ==> gamma_load32(Gamma_mem, $x_addr));
}

procedure {:extern} guarantee_reflexive();
  modifies Gamma_mem, mem;

implementation {:extern} guarantee_reflexive()
{
  assert ((memory_load32_le(mem, $x_addr) == memory_load32_le(mem, $x_addr)) && (memory_load32_le(mem, $z_addr) == memory_load32_le(mem, $z_addr)));
}

procedure main();
  modifies CF, Gamma_CF, Gamma_NF, Gamma_R0, Gamma_R31, Gamma_R8, Gamma_VF, Gamma_ZF, Gamma_mem, Gamma_stack, NF, R0, R31, R8, VF, ZF, mem, stack;
  requires (gamma_load32(Gamma_mem, $x_addr) == true);
  free requires (memory_load64_le(mem, 69664bv64) == 0bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free requires (memory_load8_le(mem, 1964bv64) == 1bv8);
  free requires (memory_load8_le(mem, 1965bv64) == 0bv8);
  free requires (memory_load8_le(mem, 1966bv64) == 2bv8);
  free requires (memory_load8_le(mem, 1967bv64) == 0bv8);
  free requires (memory_load64_le(mem, 69576bv64) == 69684bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free requires (memory_load64_le(mem, 69056bv64) == 1872bv64);
  free requires (memory_load64_le(mem, 69064bv64) == 1792bv64);
  free requires (memory_load64_le(mem, 69592bv64) == 1876bv64);
  free ensures (Gamma_R31 == old(Gamma_R31));
  free ensures (R31 == old(R31));
  free ensures (memory_load8_le(mem, 1964bv64) == 1bv8);
  free ensures (memory_load8_le(mem, 1965bv64) == 0bv8);
  free ensures (memory_load8_le(mem, 1966bv64) == 2bv8);
  free ensures (memory_load8_le(mem, 1967bv64) == 0bv8);
  free ensures (memory_load64_le(mem, 69576bv64) == 69684bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free ensures (memory_load64_le(mem, 69056bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 69064bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1876bv64);

implementation main()
{
  var #4: bv32;
  var Gamma_#4: bool;
  lmain:
    assume {:captureState "lmain"} true;
    R31, Gamma_R31 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), 0bv32), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), true);
    assume {:captureState "%000002f9"} true;
    R8, Gamma_R8 := 65536bv64, true;
    call rely();
    R8, Gamma_R8 := memory_load64_le(mem, bvadd64(R8, 4040bv64)), (gamma_load64(Gamma_mem, bvadd64(R8, 4040bv64)) || L(mem, bvadd64(R8, 4040bv64)));
    call rely();
    R8, Gamma_R8 := zero_extend32_32(memory_load32_le(mem, R8)), (gamma_load32(Gamma_mem, R8) || L(mem, R8));
    #4, Gamma_#4 := bvadd32(R8[32:0], 4294967295bv32), Gamma_R8;
    VF, Gamma_VF := bvnot1(bvcomp33(sign_extend1_32(bvadd32(#4, 1bv32)), bvadd33(sign_extend1_32(R8[32:0]), 0bv33))), (Gamma_R8 && Gamma_#4);
    CF, Gamma_CF := bvnot1(bvcomp33(zero_extend1_32(bvadd32(#4, 1bv32)), bvadd33(zero_extend1_32(R8[32:0]), 4294967296bv33))), (Gamma_R8 && Gamma_#4);
    ZF, Gamma_ZF := bvcomp32(bvadd32(#4, 1bv32), 0bv32), Gamma_#4;
    NF, Gamma_NF := bvadd32(#4, 1bv32)[32:31], Gamma_#4;
    R8, Gamma_R8 := zero_extend32_32(bvadd32(#4, 1bv32)), Gamma_#4;
    assert Gamma_ZF;
    goto lmain_goto_l00000330, lmain_goto_l0000032d;
  l00000330:
    assume {:captureState "l00000330"} true;
    R8, Gamma_R8 := 1bv64, true;
    goto l00000333;
  l0000032d:
    assume {:captureState "l0000032d"} true;
    R8, Gamma_R8 := 0bv64, true;
    goto l00000333;
  l00000333:
    assume {:captureState "l00000333"} true;
    assert Gamma_R8;
    goto l00000333_goto_l0000033b, l00000333_goto_l00000363;
  l0000033b:
    assume {:captureState "l0000033b"} true;
    R8, Gamma_R8 := 1bv64, true;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R8[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R8);
    assume {:captureState "%0000034b"} true;
    goto l0000034e;
  l00000363:
    assume {:captureState "l00000363"} true;
    goto l00000364;
  l00000364:
    assume {:captureState "l00000364"} true;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), 0bv32), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), true);
    assume {:captureState "%0000036c"} true;
    goto l0000034e;
  l0000034e:
    assume {:captureState "l0000034e"} true;
    R0, Gamma_R0 := zero_extend32_32(memory_load32_le(stack, bvadd64(R31, 12bv64))), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    goto main_return;
  lmain_goto_l0000032d:
    assume {:captureState "lmain_goto_l0000032d"} true;
    assume (bvnot1(bvcomp1(ZF, 1bv1)) != 0bv1);
    goto l0000032d;
  lmain_goto_l00000330:
    assume {:captureState "lmain_goto_l00000330"} true;
    assume (bvnot1(bvcomp1(ZF, 1bv1)) == 0bv1);
    goto l00000330;
  l00000333_goto_l0000033b:
    assume {:captureState "l00000333_goto_l0000033b"} true;
    assume (bvcomp1(R8[1:0], 1bv1) != 0bv1);
    goto l0000033b;
  l00000333_goto_l00000363:
    assume {:captureState "l00000333_goto_l00000363"} true;
    assume (bvcomp1(R8[1:0], 1bv1) == 0bv1);
    goto l00000363;
  main_return:
    assume {:captureState "main_return"} true;
    return;
}

